zookeeper/src/main/java/com/linecorp/armeria/server/zookeeper/ServerZKConnector.java

```diff
@@ -0,0 +1,109 @@
+
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.ZooDefs.Ids;
+import org.apache.zookeeper.ZooKeeper;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.zookeeper.DefaultNodeValueCodec;
+import com.linecorp.armeria.common.zookeeper.NodeValueCodec;
+import com.linecorp.armeria.common.zookeeper.ZKConnector;
+import com.linecorp.armeria.common.zookeeper.ZooKeeperException;
+
+/**
+ * A Server connection maintains the underlying connection and hearing notice from a ZooKeeper cluster.
+ */
+public class ServerZKConnector extends ZKConnector {
+    private final Endpoint endpoint;
+    private final NodeValueCodec nodeValueCodec;
+
+    /**
+     * Create a server connector.
+     * @param zkConnectionStr ZooKeeper connection string
+     * @param zNodePath       ZooKeeper node path(under which this server will registered)
+     * @param sessionTimeout  session timeout
+     * @param endpoint        register information
+     * @param nodeValueCodec           nodeValueCodec used
+     */
+    public ServerZKConnector(String zkConnectionStr, String zNodePath, int sessionTimeout, Endpoint endpoint,
+                             NodeValueCodec nodeValueCodec) {
+        super(zkConnectionStr, zNodePath, sessionTimeout);
+        this.endpoint = requireNonNull(endpoint, "endpoint");
+        this.nodeValueCodec = requireNonNull(nodeValueCodec, "nodeValueCodec");
+    }
+
+    /**
+     * Create a server connector.
+     * @param zkConnectionStr ZooKeeper connection string
+     * @param zNodePath       ZooKeeper node path(under which this server will registered)
+     * @param sessionTimeout  session timeout
+     * @param endpoint        register information
+     */
+    public ServerZKConnector(String zkConnectionStr, String zNodePath, int sessionTimeout, Endpoint endpoint) {
+        this(zkConnectionStr, zNodePath, sessionTimeout, endpoint, new DefaultNodeValueCodec());
+        connect();
+    }
+
+    @Override
+    protected void postConnected(ZooKeeper zooKeeper) {
+        try {
+            //first check the parent node
+            if (zooKeeper.exists(getzNodePath(), false) == null) {
+                //parent node not exist, create it
+                try {
+                    zooKeeper.create(getzNodePath(), endpoint.host().getBytes(StandardCharsets.UTF_8),
+                                     Ids.OPEN_ACL_UNSAFE,
+                                     CreateMode.PERSISTENT);
+                } catch (KeeperException exception) {
+                    //other server has created the node in concurrently
+                }
+            }
+            checkNode(zooKeeper);
+        } catch (Exception ex) {
+            throw new ZooKeeperException(ex);
+        }
+    }
+
+    @Override
+    protected void nodeChange(ZooKeeper zooKeeper, String path) {
+        if (path.equals(getzNodePath() + '/' + endpoint.host() + '_' + endpoint.port())) {
```

Comment:
no , we don't. this change event was triggered by ZooKeeper cluster, it  must have path attribute, will not be null